### PR TITLE
[Pipelining] Support separate dI / dW and V-schedules

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139415
* #138886
* #138924
* __->__ #131762

### Separate dI / dW:

PipelineScheduleRuntime now supports execution of merged FULL_BACKWARD
or separate dI / dW operations.

Separating the B and W may add execution overhead or may be suboptimal
in cases where BW are 'fused', but it is worthwhile when separating B, W
lets the schedule be more efficient by filling in bubbles.  In some
cases, the schedule will still issue B followed by W at certain points,
so in these cases just merge them back into BW ops and execute them as
full backwards rather than executing a B followed by a W.

### V-schedules:

V-schedules have a special case where the last rank has 2 adjacent
stages.

E.g. if rank3 had stage 3 and stage 4, then we should implement direct
transfer of stage3 outputs to stage4 inputs without a
send/recv.

In the schedling logic, we also must allow scheduling the
stage 4 forward after running stage 3 forward, without expecting a stage
4 RECV_F

In the runtime, we pass activations between adjacent stages without
using SEND/RECV ops since the stages are on the same rank/process.  We
add new APIs to PipelineStage abstraction for passing the activations
both during forward and backward.  Currently the implementation directly
modifies the 'recv buffers' the stage is managing, so the
forward/backwrad execution logic does not need to know the difference.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/pipelining/schedules.py [hunk 1]
```diff
@@ -2218,7 +2263,7 @@ def _ready_to_schedule(action: Optional[_Action]) -> bool:
                 ):
                     return True
             return False
-        elif action.computation_type == B:
+        elif action.computation_type in (B, BW):
```
Comment (H-Huang) [inline torch/distributed/pipelining/schedules.py:2266]: From what I could tell, all the `BW` changes are changes the the schedule plan side, but don't see need the logic where `BW` actually gets executed? As in where is the logic that checks:
```python
if computation_type == "BW":
    # perform backward_one_chunk(), etc.
```

File: torch/distributed/pipelining/schedules.py [hunk 2]
```diff
@@ -94,6 +96,8 @@ def from_str(action):
             return _ComputationType.SEND_B
         elif action == "RECV_B":
             return _ComputationType.RECV_B
+        elif action == "BW":
```
Comment (H-Huang) [inline torch/distributed/pipelining/schedules.py:99]: For weight-deferred schedules (e.g. interleaved1f1b), `B` stands for a "full backward", while in zero-bubble `B` stands for "backward for inputs" and `BW` actually stands for the full backward. So maybe for better consistency actually the non-zerobubble schedules should be using `BW` in their schedules instead of `B`? Or should we just give a note about this to the users.

Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:99]: Yes, if you are good with the current way (BW=FULL_BACKWARD) I can update all the other schedules in this PR.  I also need to add the execution part as you pointed out.

Comment (H-Huang) [reply torch/distributed/pipelining/schedules.py:99]: Alternatively we could keep B=FULL_BACKWARD and add a new I=INPUT for input only? I'm fine with either

File: test/distributed/pipelining/schedule_registry.py [hunk 3]
```diff
@@ -139,6 +140,8 @@ def __init__(
         self.use_full_backward = False
 
         # Go through two microbatches
+        # TODO(whc) unify the semantics of the IR for old runtime with new runtime.
```
Comment (wconstab) [inline test/distributed/pipelining/schedule_registry.py:143]: fixed in later PR in this stack

File: torch/distributed/pipelining/schedules.py [hunk 4]
```diff
@@ -1289,15 +1335,19 @@ def _step_microbatches(
                         )
                         self._maybe_compute_loss(stage, output, target_mbs, mb_index)
                         ops.extend(stage.get_fwd_send_ops(mb_index))
-                    elif computation_type == _ComputationType.BACKWARD:
+
+                    # TODO(whc) for now i'm going with the hopefully backward-compatible position that legacy IR with
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:1339]: fixed in later PR in this stack

File: torch/distributed/pipelining/schedules.py [hunk 5]
```diff
@@ -2213,7 +2295,7 @@ def _ready_to_schedule(action: Optional[_Action]) -> bool:
                 ):
                     return True
             return False
-        elif action.computation_type == B:
+        elif action.computation_type in (BACKWARD_INPUT, FULL_BACKWARD):
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:2298]: I think I should change this to just 'FULL_BACKWARD' to be consistent with the rest of this PR and then add BACKWARD_INPUT back in in the later PR where I fix other inconsistencies.

File: test/distributed/pipelining/artifacts/zb1p_2rank_2stagep_compute.csv [hunk 6]
```diff
@@ -0,0 +1,2 @@
+0F0,0F1,2F0,,2F1,2I0,2W0,0F2,2I1,2W1,0F3,0I0,0W0,2F2,0I1,0W1,2F3,2I2,2W2,0F4,2I3,2W3,0F5,0I2,0W2,2F4,0I3,0W3,2F5,2I4,2W4,0F6,2I5,2W5,0F7,0I4,0W4,2F6,0I5,0W5,2F7,2I6,2W6,2I7,2W7,0I6,0W6,0I7,0W7
```
Comment (H-Huang) [inline test/distributed/pipelining/artifacts/zb1p_2rank_2stagep_compute.csv:1]: For my understanding, if you load the compute csv without the comms csv will it error? Or will it automatically determine the comms for you

**Review by H-Huang — APPROVED** [review]

File: torch/distributed/pipelining/schedules.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py]: fixed

File: test/distributed/pipelining/artifacts/zb1p_2rank_2stagep_compute.csv (hunk 6 above)
Comment (wconstab) [reply test/distributed/pipelining/artifacts/zb1p_2rank_2stagep_compute.csv:1]: within test_csv, it is explicit: we load the compute-only one, run the lowering passes, then compare the output of that with the saved comms one.

for real users, it is API'd somewhat: load_csv in PipelineScheduleRuntime accepts a kwarg for whether its a comms or compute csv.  If its a compute one, then it will run add_send_recv.

I think before we more widely roll this out, we should better define the api around the lowering passes.  Probably a function for 'lowering' the schedule and some config flags for any optional passes.  For now its kind of manual: load_csv just hardcodes which passes to run.

File: torch/distributed/pipelining/stage.py [hunk 7]
```diff
@@ -301,6 +334,84 @@ def _get_recv_ops(
...
+            if not isinstance(info, _RecvInfo):
+                # TODO: when would info not be a _RecvInfo? should this be an error?
+                continue
+
+            # We don't need to do a data copy here, since we can directly pass the activation tensor reference from
+            # one stage to the next.  However, we do need to mark the activation as a leaf tensor since it will serve
+            # as the input tensor for a fresh autograd graph, not part of the previous stage's autograd graph.
+            # TODO: confirm, do we use this activation as the root of the backward call for the previous stage? does
+            # detach have any affect on that?
+            info.buffer = tensor.detach().requires_grad_(True)
```
Comment (wconstab) [inline torch/distributed/pipelining/stage.py:376]: @H-Huang i should remove this TODO.  It seems to pass the gradient tests but I wonder if you have any more insight, if I am doing the best thing here.

File: torch/distributed/pipelining/stage.py [hunk 8]
```diff
@@ -301,6 +334,84 @@ def _get_recv_ops(
...
+        """
+        Returns the input grad tensors for this stage, which correspond to the stage inputs during forward.
+        """
+        assert (
+            self.has_backward
+        ), "can't steal_bwd_input if this stage doesn't have backward"
+        assert not self.is_first, "can't get bwd output if this stage is first"
+
+        self._check_chunk_id(mb_index)
+        # TODO(whc) we should be indexing mb_index into self.grads_input, but it appears we are only storing
```
Comment (wconstab) [inline torch/distributed/pipelining/stage.py:388]: made an issue
https://github.com/pytorch/pytorch/issues/139404

File: torch/distributed/pipelining/stage.py [hunk 9]
```diff
@@ -301,6 +334,84 @@ def _get_recv_ops(
...
+        # the most recently created grads which needs to be fixed not only here but also for get_bwd_send_ops.
+
+        return self.grads_input
+
+    def set_local_bwd_input(self, next_stage_bwd_outputs: Tuple[Any], mb_index: int):
+        """
+        Moves 'grad input' tensors from the next stage to 'grad_output' on this stage, avoiding a copy or send/recv.
+        Does not detach or set '_requires_grad'.
+        """
+        # TODO(whc) discrepancy between list/tuple type here. need to clean up
```
Comment (wconstab) [inline torch/distributed/pipelining/stage.py:398]: i think this TODO should be removed as we decided that it is expected that users could choose tuple or tensor as return value.  the 'normalize' function should be fixing this for us now?  (but we aren't calling it?)

File: torch/distributed/pipelining/stage.py [hunk 10]
```diff
@@ -28,6 +28,39 @@
...
+    The loss function can be written in any way, such that its inputs match the outputs of the model.
+
+    It would be convenient if we could strictly type the output signature of the pipeline stage wrapping the model,
+    but we do not want to impose an unnecessary constraint on user provided models.
+
+    Currently, we let user provided models return either a Tensor or a tuple of Tensors from each stage. Due to
+    torch.export tracing, compiled models may also return a list instead of a Tuple, which we will normalize back to a
+    tuple for consistency.
+
+    TODO: should we be stricter about asserting that stage modules (intermediate and output) all return only Tensor
```
Comment (wconstab) [inline torch/distributed/pipelining/stage.py:50]: @H-Huang wdyt, should we be asserting this?  if a stage returned a non-tensor, we'd fail on send/recv right?

File: torch/distributed/pipelining/stage.py (hunk 9 above)
Comment (wconstab) [reply torch/distributed/pipelining/stage.py:398]: https://github.com/pytorch/pytorch/issues/139405

made an issue. looks like we currently don't know what type this will be and aren't normalizing it. lets fix in another PR.

File: torch/distributed/pipelining/stage.py [hunk 11]
```diff
@@ -301,6 +334,84 @@ def _get_recv_ops(
...
+
+        # See [Note: pipeline model output type]
+        prev_stage_outputs = _normalize_model_output_as_tuple(prev_stage_outputs)
+
+        for info, tensor in zip(recv_infos, prev_stage_outputs):
+            assert isinstance(
+                tensor, torch.Tensor
+            ), f"expected tensor values as outputs from prev stage, got {type(tensor)}"
+            if not isinstance(info, _RecvInfo):
+                # TODO: when would info not be a _RecvInfo? should this be an error?
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/pipelining/stage.py]: fixed

File: test/distributed/pipelining/test_schedule.py [hunk 12]
```diff
@@ -427,7 +560,297 @@ def test_send_recv(self, test_info):
...
+        # Create a pipeline stage to wrap that submodule
+        num_microbatches = 2
+        input_args = x.chunk(num_microbatches)[0]
+        stages = [
+            PipelineStage(
+                stage_module,
+                stage_idx,
+                n_stages,
+                device,
+                # TODO(whc) shape inference shouldn't have needed to run communications in this 1-rank, 2-stage scenario,
```
Comment (wconstab) [inline test/distributed/pipelining/test_schedule.py:672]: removed this, it is fixed now. probably fixed by @H-Huang's PR to fix single-stage schedule usage.
